pkg/controllers/consolidation/stabilizationwindow.go

```diff
@@ -0,0 +1,201 @@
+	clock clock.Clock
+
+	lastConsolidationTime time.Time
+	cluster               *state.Cluster
+	kubeClient            client.Client
+}
+
+// fastWindowTime is the  window used if all the cluster workloads are reporting ready
+const fastWindowTime = 15 * time.Second
+
+// standardWindowTime is the window used if workloads aren't ready, to allow us to make progress in the event of
+// persistently unready workloads (e.g. a bad deployment)
+const standardWindowTime = 3 * time.Minute
+
+// maxWindowTime is the maximum window time to ensure that even if we are in constant flux with pods being added &
+// removed, which either cause deployments to go unready or are in the presence of a persistently unready deployment,
+// we will still make some progress
+const maxWindowTime = 10 * time.Minute
+
+// NewStabilizationWindow maintains a dynamic stabilization window.
+// Node/Pod additions/deletions extend the window by standardWindowTime, up to the maxWindowTime
+// If there are no pending pods, deployments/replicasets are ready, etc. fastWindowTime is used instead of standardWindowTime
+//
+// t = 0, we consolidated, fast window point is at t=15s, standard is t=3min, max is=10min
+// t = 10s, a pod is deleted, fast window point is t=25s, standard is t=3m10s, max is t=10m
+// t = 25s, if the conditions for fast window are true, we consolidate.  Assume not, we won't consolidate until 3m25s.
+// t = 30s, a pod is added, fast window is t=45s, standard is t=3m30s, max is t=10m
+// assume a pod is added and deleted every few seconds which constantly resets the window
+// t = 10m, a pod is added, fast window is t=10m15s, standard is t=13m, max is still t=10m since we haven't
+// consolidated, so we consolidate now even though the cluster is in flux since we haven't done so in 10minutes.
+func NewStabilizationWindow(clk clock.Clock, cluster *state.Cluster, kubeClient client.Client) *StabilizationWindow {
+	return &StabilizationWindow{
+		clock:      clk,
+		cluster:    cluster,
+		kubeClient: kubeClient,
+	}
+}
+
+// IsStabilized returns true if we are out of the controller's stabilization window
+func (w *StabilizationWindow) IsStabilized(ctx context.Context) bool {
+	stabilizationTime := w.clock.Now().Add(-w.stabilizationWindow(ctx))
+
+	// has the cluster been unchanged for the desired amount of time?
+	clusterStable := w.cluster.LastNodeDeletionTime().Before(stabilizationTime) &&
+		w.cluster.LastNodeCreationTime().Before(stabilizationTime) &&
+		w.cluster.LastPodDeletionTime().Before(stabilizationTime) &&
+		w.cluster.LastPodCreationTime().Before(stabilizationTime)
+
+	// or has our stabilization window exceeded its max length?
+	exceededMaxStabWindow := w.lastConsolidationTime.Add(maxWindowTime).Before(w.clock.Now())
+
+	return clusterStable || exceededMaxStabWindow
+}
+
+// Reset resets the controller's stabilization window, we must get out of this window again before
+// IsStabilized will return true
+func (w *StabilizationWindow) Reset() {
```

Comment:
Thoughts on setting this in IsStabilized, so callers don't need to think about it?